content/posts/a-subtle-change-in-pdf-output.md

```diff
@@ -0,0 +1,79 @@
+---
+title: A subtle change in PDF output
+date: 2021-07-05
+language: en
+author: Dimiter Petrov
+summary: How a fallback font was breaking an integration test
+---
+
+## The problem
+
+I had dockerized a Rails application, but a test was failing when run within the Docker container. The relevant code is:
+
+```ruby
+pdf_string = PdfGenerator.new(input).generate
+pdf_pages = parse_pdf(pdf_string)
+
+expect(pdf_pages[2]).to include(
+  'A rather long line from the PDF that was in the original test, ' \
+  'but which I have substituted for the purposes of this example'
+)
+```
+
+It renders a PDF, then parses it with [pdf-reader](https://github.com/yob/pdf-reader) and asserts that some text is on page 3 of the document.
+
+The expectation failed, but the diff showed that the text *was* on the page. However it was on two lines instead of one.
+
+Sure, one could put a line break in the test and call it a day:
+
+```ruby
+expect(pdf_pages[2]).to include(
+  "A rather long line from the PDF that was in the original test, " \
+  "but which I have substituted for the purposes of this\nexample" # Note the \n
+)
+```
+
+I didn't give up so easily.
+
+## The motivation
+
+As an agency, we have our set of practices and documentation about running software. That makes provisioning, monitoring, backups, knowledge sharing, etc. easier. The application in question was an exception. It had been built by someone else and had been in production for years before Simplificator took over development and maintenance. We wanted to move it to our infrastructure eventually, but we had decided to get familiar with the code first.
+
+After a few months of developing new features and fixing bugs, we knew enough to migrate to a new server with more confidence. Running the application in Docker was one step of the process and this is when I ran into the test failure above.
+
+I did not cheat and change the test, because even though it looked like an insignificant detail, it *could* be something more. Whenever you make an infrastructure change you find things that were assumed, but undocumented, things that were done once manually and forgotten. We had written the PDF generation test in the phase when were getting to know the codebase. Now this test was helping us consistently observe a difference in behaviour between two environments (Docker and no Docker). A rare chance!
```

Comment:
```suggestion
I did not cheat and change the test, because even though it looked like an insignificant detail, it *could* be something more. Whenever you make an infrastructure change you find things that were assumed, but undocumented, things that were done once manually and forgotten. We had written the PDF generation test in the phase when we were getting to know the codebase. Now this test was helping us consistently observe a difference in behaviour between two environments (Docker and no Docker). A rare chance!
```